### PR TITLE
[ALLUXIO-2639] - List all pinned files in Alluxio shell

### PR DESCRIPTION
**Pull request by ShailShah**
https://alluxio.atlassian.net/browse/ALLUXIO-2639

### CODE REVIEW

File: shell/src/main/java/alluxio/shell/command/LsCommand.java [hunk 1]
```diff
@@ -173,8 +193,9 @@ public String getUsage() {
   @Override
   public String getDescription() {
     return "Displays information for all files and directories directly under the specified path."
-        + " Specify -R to display files and directories recursively."
-        + " Specify -d to list directories as plain files."
-        + " Specify -f to force loading files in the directory.";
+            + " Specify -R to display files and directories recursively."
```
Comment (uronce-cc) [inline shell/src/main/java/alluxio/shell/command/LsCommand.java]: please fix the indentation

File: shell/src/main/java/alluxio/shell/command/LsCommand.java [hunk 2]
```diff
@@ -128,20 +137,31 @@ private void ls(AlluxioURI path, boolean recursive, boolean forceLoadMetadata, b
...
-            forceLoadMetadata, false);
+                forceLoadMetadata, false, pinnedMetadata);
       }
     }
   }
 
   private List<URIStatus> listStatusSortedByIncreasingCreationTime(AlluxioURI path,
-      ListStatusOptions options)
-      throws AlluxioException, IOException {
+                                                                   ListStatusOptions options)
```
Comment (uronce-cc) [inline shell/src/main/java/alluxio/shell/command/LsCommand.java]: please fix the style

File: shell/src/main/java/alluxio/shell/command/LsCommand.java (hunk 1 above)
Comment (ShailShah) [inline shell/src/main/java/alluxio/shell/command/LsCommand.java]: @uronce-cc Have fixed the coding style. Could you please review it.

File: shell/src/main/java/alluxio/shell/command/LsCommand.java [hunk 3]
```diff
@@ -44,15 +44,15 @@
   /**
    * Formats the ls result string.
    *
-   * @param acl whether security is enabled
-   * @param isFolder whether this path is a file or a folder
-   * @param permission permission string
-   * @param userName user name
-   * @param groupName group name
-   * @param size size of the file in bytes
+   * @param acl          whether security is enabled
```
Comment (uronce-cc) [inline shell/src/main/java/alluxio/shell/command/LsCommand.java]: please keep the original style, no extra spaces are needed to be aligned, apply to the whole PR.

File: shell/src/main/java/alluxio/shell/command/LsCommand.java [hunk 4]
```diff
@@ -99,26 +99,33 @@ protected Options getOptions() {
...
     if (dirAsFile) {
       URIStatus status = mFileSystem.getStatus(path);
+      boolean isPinned = status.isPinned();
+      if (pinnedMetadata) {
+        if (!isPinned) {
+          return;
+        }
+      }
       System.out.print(formatLsString(SecurityUtils.isSecurityEnabled(), status.isFolder(),
-          FormatUtils.formatMode((short) status.getMode(), status.isFolder()), status.getOwner(),
```
Comment (uronce-cc) [inline shell/src/main/java/alluxio/shell/command/LsCommand.java]: keep the original style

File: shell/src/main/java/alluxio/shell/command/LsCommand.java [hunk 5]
```diff
@@ -175,6 +192,7 @@ public String getDescription() {
     return "Displays information for all files and directories directly under the specified path."
         + " Specify -R to display files and directories recursively."
         + " Specify -d to list directories as plain files."
-        + " Specify -f to force loading files in the directory.";
+        + " Specify -f to force loading files in the directory."
+        + " Specify -p to list all the pinned files in the specified path";
```
Comment (uronce-cc) [inline shell/src/main/java/alluxio/shell/command/LsCommand.java]: remove `in the specified path` since the semantic is already expressed above

File: shell/src/main/java/alluxio/shell/command/LsCommand.java [hunk 6]
```diff
@@ -99,26 +99,33 @@ protected Options getOptions() {
...
+   * @param path      The {@link AlluxioURI} path as the input of the command
    * @param recursive Whether list the path recursively
    * @param dirAsFile list the directory status as a plain file
    * @throws AlluxioException when Alluxio exception occurs
-   * @throws IOException when non-Alluxio exception occurs
+   * @throws IOException      when non-Alluxio exception occurs
    */
-  private void ls(AlluxioURI path, boolean recursive, boolean forceLoadMetadata, boolean dirAsFile)
-      throws AlluxioException, IOException {
+  private void ls(AlluxioURI path, boolean recursive, boolean forceLoadMetadata, boolean dirAsFile,
```
Comment (uronce-cc) [inline shell/src/main/java/alluxio/shell/command/LsCommand.java]: it will be more maintainable to move `recursive`, `forceLoadMetadata`, `dirAsFile`, `pinnedMetadata` to `ListStatusOptions`. also s/pinnedMetadata/pinned, it's simpler and clear.

File: shell/src/main/java/alluxio/shell/command/LsCommand.java [hunk 7]
```diff
@@ -128,20 +135,30 @@ private void ls(AlluxioURI path, boolean recursive, boolean forceLoadMetadata, b
     }
     List<URIStatus> statuses = listStatusSortedByIncreasingCreationTime(path, options);
     for (URIStatus status : statuses) {
-      System.out.print(formatLsString(SecurityUtils.isSecurityEnabled(), status.isFolder(),
-          FormatUtils.formatMode((short) status.getMode(), status.isFolder()), status.getOwner(),
-          status.getGroup(), status.getLength(), status.getCreationTimeMs(),
-          100 == status.getInMemoryPercentage(), status.getPath()));
+      if (pinnedMetadata) {
+        if (status.isPinned()) {
+          System.out.print(formatLsString(SecurityUtils.isSecurityEnabled(), status.isFolder(),
```
Comment (uronce-cc) [inline shell/src/main/java/alluxio/shell/command/LsCommand.java]: this block is the same as the `else` block, and appears in other places, could you consider abstracting it out as a private method.

**Comment by ShailShah** [comment]
Have modified and included all the options in ListStatusOptions. But, I haven't modified ListStatusTOptions.

**Comment by yupeng9** [comment]
@uronce-cc do you have more comments?

File: core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java [hunk 8]
```diff
@@ -50,14 +60,85 @@ public LoadMetadataType getLoadMetadataType() {
...
    */
   public ListStatusOptions setLoadMetadataType(LoadMetadataType loadMetadataType) {
     mLoadMetadataType = loadMetadataType;
     return this;
   }
 
+  /**
+   * @return raw size option. It specifies whether to print raw sizes
+   */
+  public boolean ismRawSize() {
```
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java]: `isRawSize`

File: core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java [hunk 9]
```diff
@@ -50,14 +60,85 @@ public LoadMetadataType getLoadMetadataType() {
...
+   * @param loadMetadataType the loadMetadataType
    * @return the updated options
    */
   public ListStatusOptions setLoadMetadataType(LoadMetadataType loadMetadataType) {
     mLoadMetadataType = loadMetadataType;
     return this;
   }
 
+  /**
+   * @return raw size option. It specifies whether to print raw sizes
```
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java]: nit: please end sentences with a `.`

File: core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java [hunk 10]
```diff
@@ -50,14 +60,85 @@ public LoadMetadataType getLoadMetadataType() {
...
+   */
+  public boolean ismRawSize() {
+    return mRawSize;
+  }
+
+  /**
+   * @param rawSize Raw Size Option
```
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java]: replace "Raw Size Option" with "the value for the raw size flag to use"

File: core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java [hunk 11]
```diff
@@ -50,14 +60,85 @@ public LoadMetadataType getLoadMetadataType() {
...
+   */
+  public void setmRawSize(boolean rawSize) {
```
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java]: `setRawSize`

File: core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java [hunk 12]
```diff
@@ -50,14 +60,85 @@ public LoadMetadataType getLoadMetadataType() {
...
+    mRawSize = rawSize;
+  }
+
+  /**
+   * @return the recursive option. It specifies whether the path has to be traversed recursively
```
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java]: nit: please end sentences with a `.`

File: core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java [hunk 13]
```diff
@@ -50,14 +60,85 @@ public LoadMetadataType getLoadMetadataType() {
...
+   */
+  public boolean ismRecursive() {
```
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java]: `isRecursive`

File: core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java [hunk 14]
```diff
@@ -50,14 +60,85 @@ public LoadMetadataType getLoadMetadataType() {
...
+    return mRecursive;
+  }
+
+  /**
+   * @param recursive Recursive option. It specifies whether to display directories recursively
```
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java]: replace "Recursive option" with "the value of the recursive flag to use"

Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java]: nit: please end sentences with a `.`

File: core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java [hunk 15]
```diff
@@ -50,14 +60,85 @@ public LoadMetadataType getLoadMetadataType() {
...
+   */
+  public void setmRecursive(boolean recursive) {
+    mRecursive = recursive;
+  }
+
+  /**
+   * @return Force Load Metadata Option. It specifies whether to force loading files in the
```
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java]: replace "Force Load Metadata Option" with "the force load metadata option"

File: core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java [hunk 16]
```diff
@@ -50,14 +60,85 @@ public LoadMetadataType getLoadMetadataType() {
...
+   * @return the recursive option. It specifies whether the path has to be traversed recursively
+   */
+  public boolean ismRecursive() {
+    return mRecursive;
+  }
+
+  /**
+   * @param recursive Recursive option. It specifies whether to display directories recursively
+   */
+  public void setmRecursive(boolean recursive) {
```
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java]: `setRecursive`

File: core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java [hunk 17]
```diff
@@ -50,14 +60,85 @@ public LoadMetadataType getLoadMetadataType() {
...
+    mRecursive = recursive;
+  }
+
+  /**
+   * @return Force Load Metadata Option. It specifies whether to force loading files in the
+   *         directory.
+   */
+  public boolean ismForceLoadMetaData() {
```
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java]: `isForceLoadMetadata`

File: core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java [hunk 18]
```diff
@@ -50,14 +60,85 @@ public LoadMetadataType getLoadMetadataType() {
...
+    return mForceLoadMetaData;
+  }
+
+  /**
+   * @param forceLoadMetaData Option to force loading files in the directory
```
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java]: `@param forceLoadMetadata the force load metadata option to use. It specifies whether to force loading files in the directory.`

File: core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java [hunk 19]
```diff
@@ -50,14 +60,85 @@ public LoadMetadataType getLoadMetadataType() {
...
+   */
+  public void setmForceLoadMetaData(boolean forceLoadMetaData) {
```
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java]: `setForceLoadMetadata(boolean forceLoadMetadata)`

File: core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java [hunk 20]
```diff
@@ -50,14 +60,85 @@ public LoadMetadataType getLoadMetadataType() {
...
+    mForceLoadMetaData = forceLoadMetaData;
+  }
+
+  /**
+   * @return Directory As File Option. It specifies whether to list directories as plain files
```
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java]: what is this option for? can you describe your use case?

File: core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java [hunk 21]
```diff
@@ -50,14 +60,85 @@ public LoadMetadataType getLoadMetadataType() {
...
+   */
+  public boolean ismDirAsFile() {
```
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java]: `isDirAsFile`

File: core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java (hunk 20 above)
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java]: `the directory as file option. It specifies whether to list directories as plain files.`

File: core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java [hunk 22]
```diff
@@ -50,14 +60,85 @@ public LoadMetadataType getLoadMetadataType() {
...
+    return mDirAsFile;
+  }
+
+  /**
+   * @param dirAsFile Option to list directories as plain files
```
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java]: `@param the value of the the directory as file flag to use. It specifies whether to list directories as plain files.`

File: core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java [hunk 23]
```diff
@@ -50,14 +60,85 @@ public LoadMetadataType getLoadMetadataType() {
...
+   */
+  public void setmDirAsFile(boolean dirAsFile) {
```
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java]: `setDirAsFile`

File: core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java [hunk 24]
```diff
@@ -50,14 +60,85 @@ public LoadMetadataType getLoadMetadataType() {
...
+    mDirAsFile = dirAsFile;
+  }
+
+  /**
+   * @return IsPinned Option. It specifies whether to list all the pinned files
+   */
+  public boolean ismPinned() {
```
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java]: change the name of this flag to `pinnedOnly` and update the setter and getter following the style suggested in my previous comments

File: shell/src/main/java/alluxio/shell/command/LsCommand.java [hunk 25]
```diff
@@ -170,12 +172,21 @@ public int compare(URIStatus status1, URIStatus status2) {
...
+    listStatusOptions.setmDirAsFile(cl.hasOption("d"));
+    listStatusOptions.setmPinned(cl.hasOption("p"));
+    listStatusOptions.setmRawSize(cl.hasOption("raw"));
+    ls(path, listStatusOptions);
   }
 
   @Override
   public String getUsage() {
-    return "ls [-R|-d|-f|-raw] <path>";
+    return "ls [-R|-d|-f|-raw|-p] <path>";
```
Comment (jsimsa) [inline shell/src/main/java/alluxio/shell/command/LsCommand.java]: keep options sorted alphabetically

File: shell/src/main/java/alluxio/shell/command/LsCommand.java (hunk 6 above)
Comment (jsimsa) [inline shell/src/main/java/alluxio/shell/command/LsCommand.java]: @uronce-cc I agree that the use of options might help here, but the `alluxio.client.file.options.ListStatusOptions` is meant to be use for options passed to Alluxio servers through `FileSystem` interface invocations ... it makes no sense to be introducing options that have no effect on the `FileSystem` interface invocations

File: core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java (hunk 20 above)
Comment (ShailShah) [inline core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java]: @jsimsa This option (the directory as file option) was available in the LsCommand before I made any changes. I don't have any use case for this option.

**Comment by ShailShah** [comment]
@jsimsa I have made the requested changes. PTAL. I got confused about the usage of ListStatusOptions class. I have removed the options from the ListStatusOptions class and reverted it back to the original state. I have added the `-p` option in the LsCommand itself instead of creating the new class.

File: core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java [hunk 26]
```diff
@@ -15,8 +15,6 @@
 import alluxio.thrift.ListStatusTOptions;
 import alluxio.wire.LoadMetadataType;
 
-import com.fasterxml.jackson.annotation.JsonInclude;
```
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/file/options/ListStatusOptions.java]: please undo these changes

File: shell/src/main/java/alluxio/shell/command/LsCommand.java [hunk 27]
```diff
@@ -117,15 +125,17 @@ protected Options getOptions() {
    * @throws IOException when non-Alluxio exception occurs
    */
   private void ls(AlluxioURI path, boolean recursive, boolean forceLoadMetadata, boolean dirAsFile,
-                  boolean rawSize)
+                  boolean rawSize, boolean pinned)
```
Comment (jsimsa) [inline shell/src/main/java/alluxio/shell/command/LsCommand.java]: rename `pinned` to `pinnedOnly`

File: shell/src/main/java/alluxio/shell/command/LsCommand.java [hunk 28]
```diff
@@ -135,21 +145,22 @@ private void ls(AlluxioURI path, boolean recursive, boolean forceLoadMetadata, b
     }
     List<URIStatus> statuses = listStatusSortedByIncreasingCreationTime(path, options);
     for (URIStatus status : statuses) {
-      System.out.print(
-          formatLsString(rawSize, SecurityUtils.isSecurityEnabled(), status.isFolder(),
-          FormatUtils.formatMode((short) status.getMode(), status.isFolder()), status.getOwner(),
-          status.getGroup(), status.getLength(), status.getCreationTimeMs(),
-          100 == status.getInMemoryPercentage(), status.getPath()));
+      if (pinned) {
```
Comment (jsimsa) [inline shell/src/main/java/alluxio/shell/command/LsCommand.java]: replace the whole if-then-else block with

```
if (!pinnedOnly || status.isPinned()) {
  printLsString(status, rawSize);
}
```

File: shell/src/main/java/alluxio/shell/command/LsCommand.java [hunk 29]
```diff
@@ -184,6 +196,7 @@ public String getDescription() {
         + " Specify -R to display files and directories recursively."
         + " Specify -d to list directories as plain files."
         + " Specify -f to force loading files in the directory."
-        + " Specify -raw to print raw sizes.";
+        + " Specify -raw to print raw sizes."
```
Comment (jsimsa) [inline shell/src/main/java/alluxio/shell/command/LsCommand.java]: keep options sorted alphabetically

**Comment by ShailShah** [comment]
@jsimsa Made the required changes, PTAL.

**Comment by ShailShah** [comment]
@jsimsa Have added the test for `-p` functionality. PTAL.

**Comment by jsimsa** [comment]
Jenkins, test this please

**Comment by jsimsa** [comment]
ok to test

**Comment by jsimsa** [comment]
@ShailShah thank you for your contribution
